plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java

```diff
@@ -816,39 +949,69 @@ public Iterator<TableColumnsMetadata> streamTableColumns(ConnectorSession sessio
+
+                        String tableLocation = HiveMetastoreBackedDeltaLakeMetastore.getTableLocation(table);
+                        if (cacheTableMetadata && table.getParameters().containsKey(TRINO_LAST_TRANSACTION_VERSION) && table.getParameters().containsKey(TRINO_METADATA_SCHEMA_STRING)) {
+                            long version = Long.parseLong(table.getParameters().get(TRINO_LAST_TRANSACTION_VERSION));
+                            if (isLatestVersion(fileSystem, tableLocation, version)) {
+                                List<ColumnMetadata> columnsMetadata = DeltaLakeSchemaSupport.getColumnsMetadata(typeManager, table.getParameters().get(TRINO_METADATA_SCHEMA_STRING));
+                                return Stream.of(TableColumnsMetadata.forTable(tableName, columnsMetadata));
+                            }
+                        }
+                        // Don't store cache in streamTableColumns method for avoiding too many update calls
+
+                        TableSnapshot snapshot = transactionLogAccess.loadSnapshot(session, tableName, tableLocation, Optional.empty());
                         MetadataEntry metadata = transactionLogAccess.getMetadataEntry(session, snapshot);
                         ProtocolEntry protocol = transactionLogAccess.getProtocolEntry(session, snapshot);
                         List<ColumnMetadata> columnMetadata = getTableColumnMetadata(metadata, protocol);
-                        return Stream.of(TableColumnsMetadata.forTable(table, columnMetadata));
+                        return Stream.of(TableColumnsMetadata.forTable(tableName, columnMetadata));
                     }
                     catch (NotADeltaLakeTableException | IOException e) {
                         return Stream.empty();
                     }
                     catch (RuntimeException e) {
                         // this may happen when table is being deleted concurrently, it still exists in metastore but TL is no longer present
                         // there can be several different exceptions thrown this is why all RTE are caught and ignored here
-                        LOG.debug(e, "Ignored exception when trying to list columns from %s", table);
+                        LOG.debug(e, "Ignored exception when trying to list columns from %s", tableName);
                         return Stream.empty();
                     }
                 })
                 .iterator();
     }
 
+    private boolean isLatestVersion(TrinoFileSystem fileSystem, String tableLocation, long version)
+    {
+        String transactionLogDir = getTransactionLogDir(tableLocation);
+        Location transactionLogJsonEntryPath = getTransactionLogJsonEntryPath(transactionLogDir, version);
+        Location nextTransactionLogJsonEntryPath = getTransactionLogJsonEntryPath(transactionLogDir, version + 1);
+        try {
+            return fileSystem.newInputFile(transactionLogJsonEntryPath).exists() &&
```

Comment:
Question, no action needed: Does S3 offer maybe a way to check existence through HEAD method for multiple keys?

Reply:
It doesn't provide such capability as far as I searched.